lib/conduit_lwt_unix.ml

```diff
@@ -225,19 +225,19 @@ module Sockaddr_server = struct
     let s = match on with
     | `Socket s -> s
     | `Sockaddr sockaddr -> init_socket sockaddr in
-    async (fun () ->
-      stop >>= fun () ->
-      cont := false;
-      return_unit
-    );
+    let stop' = Lwt.map (fun () -> `Stop) stop in
     let rec loop () =
-      if not !cont then return_unit
-      else
-        Lwt_unix.accept s >>=
-        process_accept ?timeout callback >>= fun () ->
+      let accept = Lwt_unix.accept s in
+      Lwt.choose [Lwt.map (fun v -> `Accept v) accept;
+                  stop'] >>= function
+      | `Stop ->
+        Lwt.cancel accept;
+        Lwt.return_unit
+      | `Accept v ->
+        process_accept ?timeout callback v >>= fun () ->
         loop ()
     in
-    loop ()
+    Lwt.finalize loop (fun () -> Lwt_unix.close s)
```

Comment:
Lwt.finalize is not about GC. It's just run `f` regardless of success or fail. If Lwt_unix.accept will fail for any reason `Stop code will not run.